lsp-mode.el

```diff
@@ -6204,15 +6209,64 @@ SYM can be either DocumentSymbol or SymbolInformation."
       (cdr)
       (or "Other")))
 
+(defun lsp--get-line-and-col (sym)
+  "Obtain the line and column corresponding to SYM."
+  (-let* ((location (gethash "location" sym))
+          (name-range (or (and location (gethash "range" location))
+                          (gethash "selectionRange" sym)))
+          (start (gethash "start" name-range))
+          ((&hash "line" "character") start))
+    (cons line character)))
+
+(defun lsp--collect-lines-and-cols (symbols)
+  "Return a sorted list ((line . col) ...) of the locations of SYMBOLS."
+  (let ((stack (lsp--imenu-filter-symbols symbols))
+        line-col-list)
+    (while stack
+      (let ((sym (pop stack)))
+        (push (lsp--get-line-and-col sym) line-col-list)
+        (unless (seq-empty-p (gethash "children" sym))
+          (setf stack (append (lsp--imenu-filter-symbols (gethash "children" sym)) stack)))))
+    (-sort #'lsp--line-col-comparator line-col-list)))
+
+(defun lsp--convert-line-col-to-points-batch (line-col-list)
+  "Convert a list of sorted positions from line-column
+representation to point representation."
+  (let* ((line-col-to-point-map (ht-create))
+         (inhibit-field-text-motion t)
+         (curr-line 0))
+    (save-excursion
+      (save-restriction
+        (widen)
+        (goto-char (point-min))
+        (cl-loop for (line . col) in line-col-list do
+                 (forward-line (- line curr-line))
+                 (setq curr-line line)
+                 (let ((line-end (line-end-position)))
+                   (if (or (not col) (> col (- line-end (point))))
+                       (goto-char line-end)
+                     (forward-char col)))
+                 (ht-set! line-col-to-point-map (cons line col) (if imenu-use-markers
+                                                                    (point-marker)
+                                                                  (point))))))
+    line-col-to-point-map))
+
+(defun lsp--sym-line-col-comparator (sym1 sym2)
+  (lsp--line-col-comparator (lsp--get-line-and-col sym1) (lsp--get-line-and-col sym2)))
+
+(cl-defun lsp--line-col-comparator ((l1 . c1) (l2 . c2))
+  (or (< l1 l2)
+      (and (= l1 l2) (< c1 c2))))
+
 (defun lsp--imenu-create-index ()
   "Create imenu index from document symbols."
-  (let ((symbols (lsp--get-document-symbols)))
-    (if (lsp--imenu-hierarchical-p symbols)
-        (lsp--imenu-create-hierarchical-index symbols)
-      (seq-map (lambda (nested-alist)
-                 (cons (car nested-alist)
-                       (seq-map #'lsp--symbol-to-imenu-elem (cdr nested-alist))))
-               (seq-group-by #'lsp--get-symbol-type (lsp--imenu-filter-symbols symbols))))))
+  (let* ((symbols (lsp--get-document-symbols))
+         (filtered-symbols (lsp--imenu-filter-symbols symbols))
+         (line-col-to-point-map (lsp--convert-line-col-to-points-batch (lsp--collect-lines-and-cols filtered-symbols))))
```

Comment:
nit: line-col-to-point-map seems to be redundant and event better to have `(let ((lsp--line-col-to-point-hash-table ...)` which will remove the need lsp--line-col-to-point-hash-table to be cleared explicitly. Also, consider passing it as parameter.

Reply:
OK, I used `(let ((lsp--line-col-to-point-hash-table` as you've suggested.

About passing it as a formal parameter: I initially passed it as a formal parameter to `lsp--imenu-create-hierarchical-index` and `lsp--symbol-to-hierarchical-imenu-elem`, and `lsp--imenu-create-non-hierarchical-index`. But the first two functions are mutually recursive, and I think it's confusing to have recursive functions calling themselves with the same arguments everytime.